### ISSUE TITLE
Inductor generates incorrect CPU code for `uint8` operations

### ISSUE THREAD

**Issue by ar0ck** [pytorch/pytorch#105264]
### 🐛 Describe the bug

Consider the following snippet, which adds a `torch.uint8` tensor to itself, _then_ casts the result to `torch.int16`:

```python
import torch

def f(x):
    return (x + x).to(torch.int16)

x = torch.tensor(128, dtype=torch.uint8)
print(f(x))
print(torch.compile(f)(x))
```

This produces the following output (`0` is correct):

```python
tensor(0, dtype=torch.int16)
tensor(256, dtype=torch.int16)
```

The relevant generated CPU kernel:

```cpp
extern "C" void kernel(const unsigned char* in_ptr0,
                       short* out_ptr0)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        auto tmp1 = tmp0 + tmp0;
        auto tmp2 = static_cast<short>(tmp1);
        out_ptr0[static_cast<long>(0L)] = tmp2;
    }
}
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230714+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.5 (default, Nov 23 2021, 15:27:38)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-148-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 530.30.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          32
On-line CPU(s) list:             0-31
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Silver 4314 CPU @ 2.40GHz
Stepping:                        6
Frequency boost:                 enabled
CPU MHz:                         3136.478
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4800.00
Virtualization:                  VT-x
L1d cache:                       768 KiB
L1i cache:                       512 KiB
L2 cache:                        20 MiB
L3 cache:                        24 MiB
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.25.1
[pip3] torch==2.1.0.dev20230714+cpu
[conda] No relevant packages
```

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

**Comment by davidradl** [comment]
I am going to investigate this one. Let me know if someone has already done the work

**Comment by davidradl** [comment]
the exact error I get is 
`BackendCompilerFailed: debug_wrapper raised CppCompileError: C++ compile error

Command:
g++ /var/folders/wv/pmg4pqgs0kxb2471qr3bplzh0000gn/T/torchinductor_davidradley/wy/cwycykkebzocy32ebqobzocvzulv2vjezczykngowdcijsoivmrk.cpp -shared -fPIC -Wall -std=c++17 -Wno-unused-variable -I/Users/davidradley/IdeaProjects/PythonTest/venv/lib/python3.9/site-packages/torch/include -I/Users/davidradley/IdeaProjects/PythonTest/venv/lib/python3.9/site-packages/torch/include/torch/csrc/api/include -I/Users/davidradley/IdeaProjects/PythonTest/venv/lib/python3.9/site-packages/torch/include/TH -I/Users/davidradley/IdeaProjects/PythonTest/venv/lib/python3.9/site-packages/torch/include/THC -I/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/include/python3.9 -I/Users/davidradley/miniconda3/include -L/Users/davidradley/miniconda3/lib -lomp -O3 -ffast-math -fno-finite-math-only -Xclang -fopenmp -D C10_USING_CUSTOM_GENERATED_MACROS -o/var/folders/wv/pmg4pqgs0kxb2471qr3bplzh0000gn/T/torchinductor_davidradley/wy/cwycykkebzocy32ebqobzocvzulv2vjezczykngowdcijsoivmrk.so

Output:
In file included from /var/folders/wv/pmg4pqgs0kxb2471qr3bplzh0000gn/T/torchinductor_davidradley/wy/cwycykkebzocy32ebqobzocvzulv2vjezczykngowdcijsoivmrk.cpp:2:
/var/folders/wv/pmg4pqgs0kxb2471qr3bplzh0000gn/T/torchinductor_davidradley/zt/cztcl2vp5yqlnhofzpqfficjcxgyict6e3xhfdd7sdbkipp4p44x.h:6:10: fatal error: 'omp.h' file not found
#include <omp.h>
         ^~~~~~~
1 error generated.`

**Comment by davidradl** [comment]
@ar0ck can I just check that you see this error, so I can confirm that we are seeing the same symptoms.

**Comment by davidradl** [comment]
I had [some issues](https://discuss.pytorch.org/t/omp-h-not-found-running-some-dynamo-tests/184427/2) setting up my environment, but can now recreate locally. I will continue to investigate

**Comment by davidradl** [comment]
Where I have got to:
The FX graph:
```
opcode         name    target                   args                kwargs
-------------  ------  -----------------------  ------------------  --------
placeholder    l_x_    L_x_                     ()                  {}
call_function  add     <built-in function add>  (l_x_, l_x_)        {}
call_method    to      to                       (add, torch.int16)  {}
output         output  output                   ((to,),)            {}
```


The trace shows:
```
2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG] TRACED GRAPH
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG]  ===== __compiled_fn_0 =====
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG]  <eval_with_key>.0 class GraphModule(torch.nn.Module):
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG]     def forward(self, L_x_ : torch.Tensor):
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG]         l_x_ = L_x_
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG]         
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG]         # File: /Users/davidradley/pytorch/test/test2.py:17, code: return (x + x)
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG]         add = l_x_ + l_x_;  l_x_ = None
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG]         return (add,)
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG]         
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph_code: [DEBUG] 
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph: [DEBUG] TRACED GRAPH
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph: [DEBUG]  __compiled_fn_0 <eval_with_key>.0 opcode         name    target                   args          kwargs
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph: [DEBUG] -------------  ------  -----------------------  ------------  --------
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph: [DEBUG] placeholder    l_x_    L_x_                     ()            {}
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph: [DEBUG] call_function  add     <built-in function add>  (l_x_, l_x_)  {}
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph: [DEBUG] output         output  output                   ((add,),)     {}
[2023-07-22 11:25:25,754] torch._dynamo.output_graph.__graph: [DEBUG] 
[2023-07-22 11:25:25,758] torch._dynamo.output_graph.__graph_sizes: [DEBUG] TRACED GRAPH TENSOR SIZES
[2023-07-22 11:25:25,758] torch._dynamo.output_graph.__graph_sizes: [DEBUG] ===== __compiled_fn_0 =====
[2023-07-22 11:25:25,758] torch._dynamo.output_graph.__graph_sizes: [DEBUG] l_x_: ()
[2023-07-22 11:25:25,758] torch._dynamo.output_graph.__graph_sizes: [DEBUG] add: ()
[2023-07-22 11:25:25,758] torch._dynamo.output_graph.__graph_sizes: [DEBUG] 
[2023-07-22 11:25:25,758] torch._dynamo.output_graph: [INFO] Step 2: calling compiler function inductor
[2023-07-22 11:25:25,793] torch._functorch.aot_autograd.__aot_graphs: [INFO] TRACED GRAPH
[2023-07-22 11:25:25,793] torch._functorch.aot_autograd.__aot_graphs: [INFO]  ===== Forward graph 0 =====
[2023-07-22 11:25:25,793] torch._functorch.aot_autograd.__aot_graphs: [INFO]  <eval_with_key>.4 from /Users/davidradley/pytorch/torch/fx/experimental/proxy_tensor.py:477 in wrapped class <lambda>(torch.nn.Module):
[2023-07-22 11:25:25,793] torch._functorch.aot_autograd.__aot_graphs: [INFO]     def forward(self, arg0_1: i64[]):
[2023-07-22 11:25:25,793] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /Users/davidradley/pytorch/test/test2.py:17, code: return (x + x)
[2023-07-22 11:25:25,793] torch._functorch.aot_autograd.__aot_graphs: [INFO]         add: i64[] = torch.ops.aten.add.Tensor(arg0_1, arg0_1);  arg0_1 = None
[2023-07-22 11:25:25,793] torch._functorch.aot_autograd.__aot_graphs: [INFO]         return (add,)

```

The byte codes are:
```
[2023-07-22 11:25:25,750] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST x []
[2023-07-22 11:25:25,750] torch._dynamo.symbolic_convert: [DEBUG] TRACE LOAD_FAST x [TensorVariable()]
[2023-07-22 11:25:25,750] torch._dynamo.symbolic_convert: [DEBUG] TRACE BINARY_ADD None [TensorVariable(), TensorVariable()]
[2023-07-22 11:25:25,753] torch._dynamo.symbolic_convert: [DEBUG] TRACE RETURN_VALUE None [TensorVariable()]
```

I assume the BINARY_ADD is not accounting for the UINT8 size of the parameters, it should cast (specialise) to UINT8 after the add then return. Investigating where this code should be for CPUs.

**Comment by davidradl** [comment]
I am still finding my way in the code. Any pointers to the module would be helpful. I am going on vacation returning on the 16th of August. I will carry on with this then, but if someone wants to work on this before then - feel free :-)

**Comment by davidradl** [comment]
I am back now and will continue on looking into this.

**Comment by davidradl** [comment]
A summary: the issue is for cpus with the compiler, with a simple function, to add to int8 variables together, gives the wrong value, if we add 128 to 128, which with uints should be 0 (because of the size of a uint - this happens without the compiler. The error is that it adds up to 256. the issue appears to be that the generated code just adds the tensors together, it does not honour their smaller type. I see the autograd generated code seems to cast down the function arguments. I suspect the variables are floats on the stack and need to be cast to their dtype prior to adding; I see comments saying that uint8s are stored as floats in the code. I cannot see in the code where this forward method content is generated. I think it is somewhere in [https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py ](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py ). I am looking into how to source debug step through this.

**Comment by davidradl** [comment]
I notice that with the latest code the generated forward function is different:

```
3:01:17,857] [0/0] torch._functorch.aot_autograd.__aot_graphs: [INFO] TRACED GRAPH
[2023-08-31 13:01:17,857] [0/0] torch._functorch.aot_autograd.__aot_graphs: [INFO]  ===== Forward graph 0 =====
[2023-08-31 13:01:17,857] [0/0] torch._functorch.aot_autograd.__aot_graphs: [INFO]  <eval_with_key>.4 from /Users/davidradley/pytorch/torch/fx/experimental/proxy_tensor.py:491 in wrapped class <lambda>(torch.nn.Module):
[2023-08-31 13:01:17,857] [0/0] torch._functorch.aot_autograd.__aot_graphs: [INFO]     def forward(self, arg0_1: u8[]):
[2023-08-31 13:01:17,857] [0/0] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /Users/davidradley/pytorch/test/test2.py:15, code: return (x + x).to(torch.int16)
[2023-08-31 13:01:17,857] [0/0] torch._functorch.aot_autograd.__aot_graphs: [INFO]         add: u8[] = torch.ops.aten.add.Tensor(arg0_1, arg0_1);  arg0_1 = None
[2023-08-31 13:01:17,857] [0/0] torch._functorch.aot_autograd.__aot_graphs: [INFO]         convert_element_type: i16[] = torch.ops.prims.convert_element_type.default(add, torch.int16);  add = None
[2023-08-31 13:01:17,857] [0/0] torch._functorch.aot_autograd.__aot_graphs: [INFO]         return (convert_element_type,)
```

the generated  kernel is still incorrect

```
extern "C" void kernel(const unsigned char* in_ptr0,
                       short* out_ptr0)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        auto tmp1 = tmp0 + tmp0;
        auto tmp2 = static_cast<short>(tmp1);
        out_ptr0[static_cast<long>(0L)] = tmp2;
    }
}
''')
```




But the reported issue still occurs.

**Comment by davidradl** [comment]
Generating this does not solve it.
```
    {
        auto tmp0 = static_cast<uint8_t>(in_ptr0[static_cast<uint8_t>(0L)]);
        auto tmp1 = tmp0 + tmp0;
        auto tmp2 = static_cast<short>(tmp1);
        out_ptr0[static_cast<uint8_t>(0L)] = tmp2;
    }
```

**Comment by davidradl** [comment]
the following works
```
def f(x):
    y=x+x
    return (y).to(torch.int16)
```

**Comment by davidradl** [comment]
the line with the add operation is created in common.py in the generate() method. I am unsure as to how I could introduce the cast. the below   buffer.writeline(line) is the one that writes the line, but at this stage the expression has been created.  
    ```
    if not var:
            var = self.newvar(bounds) if assignment else None
            self.cache[cache_key] = var
            if write:
                if V.kernel.current_node:
                    V.kernel.current_node.codegen_originating_info(
                        buffer, only_once=True
                    )
                if assignment:
                    line = f"{self.prefix}{var} = {expr}{self.suffix}"
                else:
                    line = f"{expr}{self.suffix}"
                buffer.writeline(line) 
```
@lezcano @ezyang I notice you have both been in this area - could either of you give me a hint as to how I could get the cast in for inline add operations please.  It seems that the add operation (and probably other ops) needs this special casting logic so should not be treated generically with the other binary operations.       
If this is required , I am not sure how to add this. If this is going to be very involved then maybe I should hand this issue over; though I am keen learn more if possible :-)

**Comment by ezyang** [comment]
cc @jansel @peterbell10

**Comment by jansel** [comment]
Ah looks like the C++ compiler is type promoting uint8_t + uint8_t into an int.

Doing the fix inside common.py is not ideal because that code is shared between C++ and Triton, and we also don't have type information for all the intermediates.

One way to do it would be to add something like
```python
@staticmethod
def add(a, b):
   return f"decltype({a})({a} + {b})"
```
to this class:
https://github.com/pytorch/pytorch/blob/621463a3e6b488b2bff04e355a1abd9a4c5bb2cd/torch/_inductor/codegen/cpp.py#L828

This will fix this case, but it will be slightly incorrect because if you do uint8 + float32 you will get a uint8. 


You may need to define a helper function like:
```cpp
template<typename A, typename B>
auto add_helper(A a, B b) { return a + b; }
template<>
uint8_t add_helper<uint8_t, uint8_t>(A a, B b) { return a + b; }
```
to override the default type promotion logic in C++.

That is kind of annoying though since it just handles the one op and one pair of dtypes.  The other option would be to add an FX pass that does something like:
```
for node in gm.graph.nodes:
   if (node.op == "call_function" 
       and getattr(node.meta.get(val), "dtype", None) == torch.uint8
       and node.meta["val"].device.type == "cpu"):
      ... insert a call to_dtype(node, torch.uint8) into the graph ...
```

Is this the only case where C++ type promotion logic differs from PyTorch?  Or are there other cases?

**Comment by jansel** [comment]
For the second fix, you could modify this function:
https://github.com/pytorch/pytorch/blob/621463a3e6b488b2bff04e355a1abd9a4c5bb2cd/torch/_inductor/lowering.py#L389-L396

To add something like:
```python
if dtype == torch.uint8:
  # ensure uint8 value gets truncated properly
  return ops.to_dtype(fn(*[load(index) for load in loaders]), dtype)
```

**Comment by lezcano** [comment]
I think that we can mix both of Jason's ideas:
We can have two functions (or maps) that, given two dtypes, it computes the output dtype of a binop in C++ and in PyTorch (pretty sure this already exists in PyTorch). Then, adding the `ops.to_dtype` call in the lowering if these two do not agree.
You may also need to remove some casts that are currently scattered around, as
```python
    @staticmethod
    def mul(a, b):
        return f"decltype({a})({a} * {b})"
```

The advantage of doing it at a lowering level is that you'll get both the non-vectorised and the vectorised path correct in one go.

**Comment by davidradl** [comment]
Thank you very much @lezcano and @jansel.

I have tested a fix in lowering and it looks like it works. I would like to understand this more. some questions I have are:

- I see the code is in an inner_fn in inner is in a function called make_pointwise. It send up calling Pointwise.create passing the inner function. Pointwise is a type of graph node in the ir (intermediate representation) . I see in ir.py that self.inner_fn() is called in various places including store. I am not sure when the inner_fn will be called, I can see it would be after code gen.   
- I wonder if we could rename inner and inner_fn to be more descriptive. Maybe opsOverride_fn or override_fn. 
- I am wondering why it is called Pointwise. I see Pointwise is `Occurring or true for each [point](https://en.wiktionary.org/wiki/point) of a given [set](https://en.wiktionary.org/wiki/set).` Conceptually, I am not sure how this relates to the intermediate representation.

**Comment by davidradl** [comment]
Hi @lezcano and @jansel,
I have made [a change](https://github.com/pytorch/pytorch/pull/108513) that fixes the reported problem; as per your recommendations.

I added test code and found that this issue also occurs for the signed types. The fix I produced fixes uint8 and int8. It should fix int32 - but it does not. I have included the failing test case where I add 2 maximum int32s then cast to int64 in the function. In this case
torch is `tensor(-2)`
dynamo is `tensor(4294967294)`

I have left the pr in draft until we agree what to do here. I have put the pr as not user facing - is this correct as it changes behaviour? 

I had considered relaxing the type check to case for all types, but I suspect there is a performance overhead - so just wanted to make the changes that are required to fix behaviour.

**Comment by jansel** [comment]
I'd expect the C++ compiler to optimize away a no-op conversion, so it should have no cost if it is already the correct type.

Maybe we need to also convert the inputs to fix the int32 code?  Would need to inspect the types in the output code to figure out what is going on.

**Comment by davidradl** [comment]
> I'd expect the C++ compiler to optimize away a no-op conversion, so it should have no cost if it is already the correct type.
> 
> Maybe we need to also convert the inputs to fix the int32 code? Would need to inspect the types in the output code to figure out what is going on.

I have raised #108520 to track this issue separately

**Comment by leslie-fang-intel** [comment]
Verified the fix with lastet PyTorch: 545d2126f6a7525cc1707d94906ea8801033b518
The output is:
```
(pytorch) $ python test_105264.py
tensor(0, dtype=torch.int16)
tensor(0, dtype=torch.int16)
```
The generated code is:
```
cpp_fused__to_copy_add_0 = async_compile.cpp('''
#include "/tmp/torchinductor_jianan/26/c26eqbkuxvn72gf7p2xujmqjcwf4bo6lxmp6rwborxnf4gldnimh.h"
extern "C" void kernel(const unsigned char* in_ptr0,
                       short* out_ptr0)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        auto tmp1 = decltype(tmp0)(tmp0 + tmp0);
        auto tmp2 = c10::convert<short>(tmp1);
        out_ptr0[static_cast<long>(0L)] = tmp2;
    }
}
''')
```

Feel free to re-open if any further question.
